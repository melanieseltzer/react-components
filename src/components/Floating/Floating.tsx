import React, { Component, ReactNode } from 'react';
import styled from 'styled-components';

import Input from '../Input';
import Label from '../Label';

export interface Props {
  /** Takes an Input and Label as children */
  children: ReactNode;
  /** Pass it any class to extend styling */
  className?: string;
}

class Floating extends Component<Props> {
  public static Input = Input;
  public static Label = Label;

  public render() {
    const { className, children } = this.props;

    const childrenWithProps = React.Children.map(
      children,
      (child: React.ReactElement<any>): ReactNode => {
        return React.cloneElement(child, {
          float: true
        });
      }
    );

    return (
      <FloatingWrapper className={className}>
        {childrenWithProps}
      </FloatingWrapper>
    );
  }
}

const FloatingWrapper = styled.div`
  position: relative;
`;

export default Floating;
