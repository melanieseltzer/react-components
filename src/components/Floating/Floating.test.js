import React from 'react';
import { render, cleanup } from 'react-testing-library';
import 'jest-styled-components';
import Floating from '.';

afterEach(cleanup);

describe('Floating component', () => {
  it('matches the snapshot', () => {
    const { container } = render(
      <Floating>
        <Floating.Input />
        <Floating.Label>Name</Floating.Label>
      </Floating>
    );
    expect(container).toMatchSnapshot();
  });

  it('should be rendered with at least one child component', () => {
    const { container } = render(
      <Floating>
        <Floating.Input />
        <Floating.Label>Name</Floating.Label>
      </Floating>
    );
    expect(container.firstChild.hasChildNodes).toBeTruthy();
  });

  it('should render with a class of test', () => {
    const { container } = render(
      <Floating className="test">
        <Floating.Input />
        <Floating.Label>Name</Floating.Label>
      </Floating>
    );
    expect(container.firstChild.classList.contains('test')).toBe(true);
  });

  it('should implicitely pass the float prop to children and render their styles', () => {
    const { getByTestId } = render(
      <Floating>
        <Floating.Input data-testid="child-input" />
        <Floating.Label data-testid="child-label">Name</Floating.Label>
      </Floating>
    );

    expect(getByTestId('child-label')).toHaveStyleRule('position', 'absolute');
    expect(getByTestId('child-label')).toHaveStyleRule('top', '0');
    expect(getByTestId('child-label')).toHaveStyleRule('left', '0');
    expect(getByTestId('child-input')).toHaveStyleRule('opacity', '0', {
      modifier: `
        :placeholder-shown:not(:focus)::placeholder
      `
    });
  });
});
