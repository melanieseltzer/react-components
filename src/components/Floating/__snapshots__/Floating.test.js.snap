// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Floating component matches the snapshot 1`] = `
.c1 {
  background: none;
  border: 0;
  font-size: 16px;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  padding: 1rem 0.5rem 0.5rem 0.5rem;
}

.c1:placeholder-shown + label {
  -webkit-transform: translate(0.5rem,0.8rem) scale(1);
  -ms-transform: translate(0.5rem,0.8rem) scale(1);
  transform: translate(0.5rem,0.8rem) scale(1);
}

.c1:placeholder-shown:not(:focus)::-webkit-input-placeholder,
.c1:placeholder-shown:not(:focus)::-webkit-input-placeholder {
  opacity: 0;
  -webkit-transition: inherit;
  transition: inherit;
}

.c1:placeholder-shown:not(:focus)::-moz-placeholder,
.c1:placeholder-shown:not(:focus)::-webkit-input-placeholder {
  opacity: 0;
  -webkit-transition: inherit;
  transition: inherit;
}

.c1:placeholder-shown:not(:focus):-ms-input-placeholder,
.c1:placeholder-shown:not(:focus)::-webkit-input-placeholder {
  opacity: 0;
  -webkit-transition: inherit;
  transition: inherit;
}

.c1:placeholder-shown:not(:focus)::placeholder,
.c1:placeholder-shown:not(:focus)::-webkit-input-placeholder {
  opacity: 0;
  -webkit-transition: inherit;
  transition: inherit;
}

.c1:focus::-webkit-input-placeholder,
.c1:focus::-webkit-input-placeholder {
  color: #626262;
  opacity: 1;
}

.c1:focus::-moz-placeholder,
.c1:focus::-webkit-input-placeholder {
  color: #626262;
  opacity: 1;
}

.c1:focus:-ms-input-placeholder,
.c1:focus::-webkit-input-placeholder {
  color: #626262;
  opacity: 1;
}

.c1:focus::placeholder,
.c1:focus::-webkit-input-placeholder {
  color: #626262;
  opacity: 1;
}

.c1:not(:placeholder-shown) + label,
.c1:focus + label {
  padding-top: 0.2rem;
  -webkit-transform-origin: left top;
  -ms-transform-origin: left top;
  transform-origin: left top;
  -webkit-transform: translate(0.5rem,0rem) scale(0.8);
  -ms-transform: translate(0.5rem,0rem) scale(0.8);
  transform: translate(0.5rem,0rem) scale(0.8);
}

.c2 {
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  color: #aaa;
  position: absolute;
  top: 0;
  left: 0;
}

.c2:hover {
  cursor: text;
}

.c0 {
  position: relative;
}

<div>
  <div
    class="c0"
  >
    <input
      color="#ccc"
      class="c1"
    />
    <label
      color="#aaa"
      class="c2"
    >
      Name
    </label>
  </div>
</div>
`;
